libp2p/protocols/kademlia/find.nim

```diff
@@ -7,115 +7,140 @@
-  if reply.msgType != MessageType.findNode:
-    raise newException(ValueError, "Unexpected reply message type: " & $reply.msgType)
+  let reply = Message.decode(await conn.readLp(MaxMsgSize)).valueOr:
+    raise newException(ValueError, "FindNode reply decode fail")
 
   return reply
 
+proc updatePeers(kad: KadDHT, peerInfos: seq[PeerInfo]) {.raises: [].} =
+  for p in peerInfos:
+    discard kad.rtable.insert(p.peerId)
+    # Nodes might return different addresses for a peer, so we append instead of replacing
+    try:
+      var existingAddresses = kad.switch.peerStore[AddressBook][p.peerId].toHashSet()
+      for a in p.addrs:
+        existingAddresses.incl(a)
+      kad.switch.peerStore[AddressBook][p.peerId] = existingAddresses.toSeq()
+    except KeyError as exc:
+      debug "Could not update shortlist", err = exc.msg
+    # TODO: add TTL to peerstore, otherwise we can spam it with junk
+
 proc findNode*(
-    kad: KadDHT, targetId: Key
+    kad: KadDHT, key: Key
 ): Future[seq[PeerId]] {.async: (raises: [CancelledError]).} =
-  ## Iteratively search for the k closest peers to a target ID.
-
-  var initialPeers = kad.rtable.findClosestPeerIds(targetId, kad.config.replication)
-  var state = LookupState.init(
-    targetId, initialPeers, kad.config.alpha, kad.config.replication,
-    kad.rtable.config.hasher,
-  )
-  var addrTable: Table[PeerId, seq[MultiAddress]]
-  for p in initialPeers:
-    addrTable[p] = kad.switch.peerStore[AddressBook][p]
-
-  while not state.done:
-    let toQuery = state.selectAlphaPeers()
-    debug "Find node queries",
-      peersToQuery = toQuery.mapIt(it.shortLog()), addressTable = addrTable
-    var pendingFutures = initTable[PeerId, Future[Message]]()
+  ## Iteratively search for the k closest peers to a key.
+
+  var state = LookupState.init(kad, key)
 
-    for peer in toQuery.filterIt(kad.switch.peerInfo.peerId != it):
-      state.markPending(peer)
-      let addrs = addrTable.getOrDefault(peer, @[])
-      if addrs.len == 0:
-        state.markFailed(peer)
-        continue
-      pendingFutures[peer] =
-        kad.sendFindNode(peer, addrs, targetId).wait(chronos.seconds(5))
+  while not state.done():
+    let toQuery = state.selectAlphaPeers()
 
-      state.activeQueries.inc
+    debug "Find node queries", peersToQuery = toQuery.mapIt(it.shortLog())
 
-    let (successfulReplies, timedOutPeers) = await waitRepliesOrTimeouts(pendingFutures)
+    let
+      rpcBatch = toQuery.mapIt(kad.switch.dispatchFindNode(it, key))
+      completedRPCBatch = await rpcBatch.collectCompleted(kad.config.timeout)
+    for (fut, peerId) in zip(rpcBatch, toQuery):
+      if fut.completed():
+        state.queried.incl(peerId)
 
-    for msg in successfulReplies:
+    for msg in completedRPCBatch:
       for peer in msg.closerPeers:
         let pid = PeerId.init(peer.id)
         if not pid.isOk:
           error "Invalid PeerId in successful reply", peerId = peer.id
           continue
```

Comment:
```suggestion
    addNewPeerAddresses(kad.switch.peerStore[AddressBook], msg.closerPeers)
```

second `for` should not exist here. it adds to complexity that could be replaced by function